packages/frontend/core/src/components/attachment-viewer/worker/worker.ts

```diff
@@ -0,0 +1,171 @@
+    case MessageOp.Open: {
+      const action = data[type];
+      if (!action?.blob) return;
+
+      doc = await viewer.openWithBlob(action.blob);
+
+      if (!doc) return;
+
+      post(MessageOp.Open);
+      break;
+    }
+
+    case MessageOp.SyncViewportInfo: {
+      const updated = data[type];
+
+      if (updated) {
+        Object.assign(viewportInfo, updated);
+      }
+
+      break;
+    }
+
+    case MessageOp.SyncDocInfo: {
+      if (!doc) return;
+
+      const updated = data[type];
+
+      if (updated) {
+        Object.assign(docInfo, updated);
+      }
+
+      const page = doc.page(0);
+
+      if (page) {
+        Object.assign(docInfo, {
+          cursor: 0,
+          total: doc.pageCount(),
+          height: Math.ceil(page.height()),
+          width: Math.ceil(page.width()),
+        });
+        page.close();
+        post(MessageOp.SyncDocInfo, docInfo);
+      }
+      break;
+    }
+
+    case MessageOp.Render: {
+      if (!doc) return;
+
+      const { index, kind } = data[type];
+
+      let imageBitmap = cached.size > 0 ? cached.get(index) : undefined;
+      if (imageBitmap) {
+        if (kind === RenderKind.Thumbnail) {
+          const rw = 94 * viewportInfo.dpi;
+          const rh = (docInfo.height / docInfo.width) * rw;
+          imageBitmap = await resizeImageBitmap(imageBitmap, {
+            resizeWidth: Math.ceil(rw),
+            resizeHeight: Math.ceil(rh),
+          });
+        }
+
+        post(MessageOp.Rendered, { index, kind, imageBitmap });
+        return;
+      }
+
+      const width = Math.ceil(docInfo.width * viewportInfo.dpi);
+      const height = Math.ceil(docInfo.height * viewportInfo.dpi);
+      const page = doc.page(index);
+
+      if (page) {
+        const bitmap = viewer.createBitmap(width, height, 0);
+        bitmap.fill(0, 0, width, height);
+        page.render(bitmap, 0, 0, width, height, 0, flags);
+
+        const data = bitmap.toBytes();
+
+        bitmap.close();
+        page.close();
+
+        if (canvas.width !== width || canvas.height !== height) {
+          canvas.width = width;
+          canvas.height = height;
+        }
+        const imageData = new ImageData(
+          new Uint8ClampedArray(data),
+          width,
+          height
+        );
+        ctx?.clearRect(0, 0, width, height);
+        ctx?.putImageData(imageData, 0, 0);
+        imageBitmap = canvas.transferToImageBitmap();
+
+        cached.set(index, imageBitmap);
+
+        if (kind === RenderKind.Thumbnail) {
+          const rw = 94 * viewportInfo.dpi;
+          const rh = (docInfo.height / docInfo.width) * rw;
+          imageBitmap = await resizeImageBitmap(imageBitmap, {
+            resizeWidth: Math.ceil(rw),
+            resizeHeight: Math.ceil(rh),
+          });
+        }
+
+        post(MessageOp.Rendered, { index, kind, imageBitmap });
+      }
+
+      break;
+    }
+  }
+}
+
+self.addEventListener('message', (event: MessageEvent<MessageData>) => {
```

Comment:
Can we make these apis atomic and stateless?

Reply:
e.g.

before:

post(Open)
post(SyncViewportInfo)
post(SyncDocInfo)
post(Render)

after:

post(OpenAndRender, {
  pdfData,
  viewportInfo,
  renderPageIndex
}) -> DocInfo

Reply:
as @pengx17 said, we can try https://github.com/Jack-Works/async-call-rpc